src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+        {
+            var left = ParseSingleCharacterClassComponent(isFirst: components.Count == 0, afterRangeMinus: false);
+            if (left.Kind == RegexKind.CharacterClassEscape ||
+                left.Kind == RegexKind.CategoryEscape ||
+                IsEscapedMinus(left))
+            {
+                // \s or \p{Lu} or \- on the left of a minus doesn't start a range. If there is a following
+                // minus, it's just treated textually.
+                components.Add(left);
+                return;
+            }
+
+            if (_currentToken.Kind == RegexKind.MinusToken && !_lexer.IsAt("]"))
+            {
+                // trivia is not allowed anywhere in a character class
+                var minusToken = ConsumeCurrentToken(allowTrivia: false);
+
+                if (_currentToken.Kind == RegexKind.OpenBracketToken)
+                {
+                    components.Add(left);
+                    components.Add(ParseCharacterClassSubtractionNode(minusToken));
+                }
+                else
+                {
+                    var right = ParseRightSideOfCharacterClassRange();
+
+                    if (TryGetRangeComponentValue(left, isRight: false, out var leftCh) &&
+                        TryGetRangeComponentValue(right, isRight: true, out var rightCh) &&
+                        leftCh > rightCh)
+                    {
+                        minusToken = minusToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                            WorkspacesResources.x_y_range_in_reverse_order,
+                            minusToken.GetSpan()));
+                    }
+
+                    components.Add(new RegexCharacterClassRangeNode(left, minusToken, right));
+                }
+            }
+            else
+            {
+                components.Add(left);
+            }
+        }
+
+        private bool IsEscapedMinus(RegexNode node)
+            => node is RegexSimpleEscapeNode simple && IsTextChar(simple.TypeToken, '-');
+
+        private bool TryGetRangeComponentValue(RegexExpressionNode component, bool isRight, out char ch)
+        {
+            // Don't bother examining the component if it has any errors already.  This also means
+            // we don't have to worry about running into invalid escape sequences and the like.
+            if (!HasProblem(component))
+            {
+                return TryGetRangeComponentValueWorker(component, out ch);
+            }
+
+            ch = default;
+            return false;
+        }
+
+        private bool TryGetRangeComponentValueWorker(RegexNode component, out char ch)
+        {
+            switch (component.Kind)
+            {
+                case RegexKind.SimpleEscape:
+                    ch = ((RegexSimpleEscapeNode)component).TypeToken.VirtualChars[0];
+                    switch (ch)
+                    {
+                        case 'a': ch = '\u0007'; break;
```

Comment:
Perhaps add comments for this characters. Is `\u0007` the bell?

Reply:
sure!

Reply:
Suggestion still applicable.

---
In reply to: [178596111](https://github.com/dotnet/roslyn/pull/23984#discussion_r178596111) [](ancestors = 178596111)

Reply:
Done.